### DISCUSSION TITLE
Eloquent API Resource Collections MissingValue

### DISCUSSION THREAD

**Discussion by Floppy012** [laravel/framework#48244] [Q&A] [answered]
Lets say I have a `UserResource` that returns `id`, `username`, `email`, `street`, `city`, `country`, `zipcode`, and so on. For showing a single user this works fine. However for my taste it's too much information for an `index` API call. I want to reduce that information to the essential stuff that is required for a simple list (in my case `id`, `username`, `email`). 

I tried solving this by creating a custom `ResourceCollection` and mapping the collection:

```php
public function toArray(Request $request): array
{
    return [
        'data' => $this->collection->map(function(ClientResource $resource) use ($request) {
            return Arr::only($resource->toArray($request), ['id', 'username', 'email']);
        }
    ];
}
```

This works fine until I try to include entries that could be `MissingValue` (i.e. some conditional entry created with `$this->when`. The `MissingValue` is no longer filtered out and is shown in the response array.

Since I'm not working that long with API Resources I'm wondering if there is a better way to solve my usecase or if this is a bug.

**Comment by Angel5a** [comment, marked as answer]
You can call to inheritance.

You have a `UserResource` which is a full resource. Create a child `BriefUserResource` with overridden `toArray()` as (fill free to make a trait):
```php
public function toArray(Request $request): array
{
    return Arr::only(parent::toArray(request), ['id', 'username', 'email']));
}
```
And instead of  `UserResource` point your `ResourceCollection` to `BriefUserResource`:
```php
public $collects = BriefUserResource::class;
```
No need to override  `toArray()` in your collection.

Next question is why do you even need to populate `street`, `city`, `country`, `zipcode`, if they wouldn't be needed? And reverse the inheritance.

**Reply by Floppy012** [reply]
That's a nice solution. Thank You.

>  if they wouldn't be needed?

The fields I provided are just for the example. In reality, there are a lot more fields that I don't need when listing users, but still to expose when getting a single user. (i.e. `GET /users` returns multiple users paginated with most important data [to show in a list in UI, `GET /users/{id}` returns a single user with all the data [for displaying a single user in UI for profile views or user editing views])
